sw/host/ot_certs/src/x509/tests.rs

```diff
@@ -0,0 +1,571 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use base64::{engine::general_purpose, Engine as _};
+use num_bigint_dig::BigUint;
+use std::collections::HashMap;
+use std::iter::zip;
+
+use foreign_types::{ForeignType, ForeignTypeRef};
+use openssl::asn1::{Asn1IntegerRef, Asn1Object, Asn1OctetStringRef, Asn1StringRef, Asn1Time};
+use openssl::ec::EcKey;
+use openssl::pkey::Public;
+use openssl::x509::{X509NameRef, X509};
+
+use crate::offsets::CertificateWithVariables;
+use crate::template::*;
+use crate::types::Type;
+use crate::x509::extension::*;
+use crate::x509::*;
+
+// Unfortunately, the rust openssl binding does not have an API to extract arbitrary extension but it exports
+// the low level function from the C library to do that.
+fn x509_get_ext_by_obj<'a>(x509: &'a X509, obj: &Asn1Object) -> Result<&'a Asn1OctetStringRef> {
+    unsafe {
+        // SAFETY the rust openssl binding guarantees that x509 and obj are valid objects.
+        let index = openssl_sys::X509_get_ext_by_OBJ(x509.as_ptr(), obj.as_ptr(), -1);
```

Comment:
It's your preference, but some styles prefer to have minimal `unsafe` blocks with the comment above it:

```suggestion
    // SAFETY the rust openssl binding guarantees that x509 and obj are valid objects.
    let index = unsafe { openssl_sys::X509_get_ext_by_OBJ(x509.as_ptr(), obj.as_ptr(), -1) };
```

The advantage is you we can use lints like [clippy::undocumented_unsafe_blocks](https://rust-lang.github.io/rust-clippy/master/#/undocumented_unsafe_blocks) and [clippy::multiple_unsafe_ops_per_block](https://rust-lang.github.io/rust-clippy/master/#/multiple_unsafe_ops_per_block) to ensure all unsafe blocks are documented.

Reply:
Yes, please use fine-grained unsafe blocks and put SAFETY comments before each instance of unsafe keyword.

Reply:
Every function call in this function is unsafe. I can split the check for index out but I am not sure there is an advantage to splitting the other three calls below.

Reply:
Since we don't have unsafe guidelines for OpenTitan (AFAICT) I think you can leave this, but to explain the reasoning:

If you just want to use `unsafe` to opt into unsafe behaviour, then this is fine. The other way to use it is to "defuse" unsafe behaviour, meaning your unsafe blocks encapsulate a "unit" of safety (which might be multiple unsafe calls). This makes it easier to audit unsafe blocks because they're only doing one unsafe thing each.

Clippy has these lints for checking `// SAFETY` comments but they only work for `unsafe` blocks, following this model.

Thinking about it again, I don't think we want to enable `clippy::multiple_unsafe_ops_per_block` since you could require multiple unsafe operations to make something safe again, but I do think we want `clippy::undocumented_unsafe_blocks` in the future, and for that to work well we should have one unsafe "thing" happening in each block.

Again, I think you can leave this right now and we'll enforce some policy for `unsafe` in the future.

Reply:
What I was trying to explain is that if I write one unsafe per call, the code will be:
```rust
let ext = unsafe {
        // SAFETY: the rust openssl binding guarantees that x509 is a valid object
        // and index is a valid integer since X509_get_ext_by_OBJ returns either an index
        // or -1.
        openssl_sys::X509_get_ext(x509.as_ptr(), index)
}
let data = unsafe {
        // SAFETY: the returned extension is a valid openssl pointer to an extension.
        openssl_sys::X509_EXTENSION_get_data(ext)
}
unsafe{
        // From the documentation of X509_EXTENSION_get_data.
        // The returned pointer is an internal value which must not be freed up.
        Asn1OctetStringRef::from_ptr(data)
    }
```
I don't think this is particularly helpful or readable, though I understand the motivation for single-op per unsafe block. I am instead thinking of those operations as `I am doing three low-level C operations and then using Asn1OctetStringRef::from_ptr to get back into the rust world`. Hopefully this makes sense?

I think it would make sense to enable `clippy::undocumented_unsafe_blocks`.

Reply:
As James said, It's a general practice that safety comment immediately preceeds the unsafe block.

Reply:
Ah yes you are right, I will move it above the unsafe block.